### PR TITLE
Use both char and byte positions

### PR DESCRIPTION
**Pull request by AntonyBlakey**

### CODE REVIEW

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/language.rs:84]: I wonder if we should compare both? for `PartialOrd` and `Ord`.
In an LSP, they can diverge because of edits to documents, while the server still holds symbols/references to both old/new tokens.
It would still be useful to compare them.

Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/language.rs:84]: it must always be the case that comparing the byte offsets and char offsets gives the same Ordering value.

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/language.rs:77]: I would use StreamPosition in that case

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/language.rs:161]: IIUC, these are also used to create the `ariadne` span, which now we send chars instead of bytes. Should we keep bytes here?

File: crates/codegen/syntax_templates/src/typescript/parser_output.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/typescript/parser_output.rs:45]: Are these duplicates?
+ another instance below.

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/language.rs:161]: ariadne is user facing, so it should be chars and not bytes

File: crates/codegen/syntax_templates/src/typescript/parser_output.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/typescript/parser_output.rs:37]: Performance of the error getter? I might be missing something.
This will only be called once for each error, only when there are errors. Is there somewhere else it can impact?

File: crates/solidity/testing/snapshots/cst_output/Expression/unicode_string_literal/generated/0.7.0.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/Expression/unicode_string_literal/generated/0.7.0.yml:9]: I suggest keeping byte offsets in tests. Since we are mainly working in Rust, it makes it easier for someone debugging tests to take a slice and inspect the string.

File: crates/codegen/syntax_templates/src/typescript/parser_output.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/typescript/parser_output.rs:37]: I wanted it to match the node position api i.e. separate char_ and byte_. I did that because it will be a common operation and otherwise it involves a lot of work to create the object across the FFI boundary.

Maybe I'm not getting your point.

File: crates/solidity/testing/snapshots/cst_output/Expression/unicode_string_literal/generated/0.7.0.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/Expression/unicode_string_literal/generated/0.7.0.yml:9]: no opinion either way, so ok.

File: crates/codegen/syntax_templates/src/typescript/parser_output.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/typescript/parser_output.rs:37]: Not blocking this PR. Let's discuss f2f in that case :)

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/language.rs:161]: `source_start` and `source_end` are used in two places:

1. In the user facing display strings: which is fine to use char offsets.
2. In constructing `Label` and `Report` offsets, which expect bytes (?).

I suspect this change can break it? I don't think we have tests that cover the second case, but we can probably add one.

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/language.rs:77]: Do we want to leak the "stream" terminology to the public API?
I don't think there is ambiguity here between two `Position`s that need differentiating with a prefix.

However, I will leave it up to you if you feel strongly about it 👍

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/language.rs:84]: 👍

**Review by OmarTawfik — CHANGES_REQUESTED** [review]
A couple of concerns about duplicate result in the getter API, and test snapshot offsets.

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/language.rs:77]: f2f: We will use `TextPosition` and `TextRange`

File: crates/codegen/syntax_templates/src/typescript/cst_types.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/typescript/cst_types.rs:49]: f2f: suggestion about exporting and using a tuple type for these return types:

```ts
type TupleType = [ foo: bigint, bar: bigint];
```

File: crates/codegen/syntax_templates/src/shared/language.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/language.rs:161]: In the label offsets it gets shown to the user, so it should also be char offsets.

**Review by OmarTawfik — APPROVED** [review]

**Pull request merged**
